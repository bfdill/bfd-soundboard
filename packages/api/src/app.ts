import * as Koa from 'koa'
import * as KoaHelmet from 'koa-helmet'
import * as KoaBody from 'koa-body'
import * as KoaRouter from 'koa-router'
import * as cors from '@koa/cors'
import * as winston from 'winston'
import { createLogger } from '@talkboy-9000/logging'
import { getHealthcheckRouter } from './controllers/healthcheck'
import { getPlayerRouter } from './controllers/player'
import { getSoundsRouter } from './controllers/sounds'
import {
  ApplicationRouter,
  ApplicationKoa,
  ApplicationState,
  IApplication,
  IApplicationContext
} from './types'
import { ISystemMiddleware, SystemMiddleware } from './middleware'

export const configureRouter = (router: ApplicationRouter) =>
  router
    .use('/healthcheck', getHealthcheckRouter().routes())
    .use('/player', getPlayerRouter().routes())
    .use('/sounds', getSoundsRouter().routes())

export const getAppLogger = () =>
  createLogger({
    app: {
      // uhh, got config
      name: 'talkboy-9000',
      version: '2.3.6'
    }
  })
export const getAppRouter = () =>
  configureRouter(new KoaRouter<ApplicationState, IApplicationContext>())

export class App implements IApplication {
  private static instance: IApplication | undefined

  constructor(
    protected readonly koa: ApplicationKoa,
    protected readonly router: ApplicationRouter,
    protected readonly logger: winston.Logger,
    protected readonly systemMiddleware: ISystemMiddleware
  ) {}

  protected setup = (
    koa: ApplicationKoa,
    router: ApplicationRouter,
    logger: winston.Logger,
    systemMiddleware: ISystemMiddleware
  ) => {
    koa.on('error', err => {
      logger.error(err)
    })

    koa
      .use(cors())
      .use(KoaHelmet())
      .use(KoaBody())
      .use(systemMiddleware.middleware)
      .use(router.routes())
      .use(router.allowedMethods())
  }

  listen = (port: number) => {
    this.logger.info({ listen: port })
    this.setup(this.koa, this.router, this.logger, this.systemMiddleware)

    return this.koa.listen(port)
  }

  static getInstance(): IApplication {
    if (this.instance !== undefined) return this.instance

    const logger = getAppLogger()

    this.instance = new App(
      new Koa<ApplicationState, IApplicationContext>(),
      getAppRouter(),
      logger,
      new SystemMiddleware(logger)
    )

    return this.instance
  }
}
