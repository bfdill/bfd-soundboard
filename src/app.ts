import * as Koa from 'koa'
import * as KoaHelmet from 'koa-helmet'
import * as KoaBody from 'koa-body'
import * as KoaRouter from 'koa-router'
import * as winston from 'winston'
import {
  createLogger,
  ILoggingMiddleware,
  LoggingMiddleware
} from './modules/logging'
import { getHealthcheckRouter } from './controllers/healthcheck'
import { getPlayerRouter } from './controllers/player'
import { soundsRouter } from './controllers/sounds'
import {
  ApplicationRouter,
  ApplicationKoa,
  ApplicationState,
  IApplicationContext,
  IApplication
} from './types'

export const configureRouter = (router: ApplicationRouter) =>
  router
    .use('/healthcheck', getHealthcheckRouter().routes())
    .use('/player', getPlayerRouter().routes())
    .use('/sounds', soundsRouter.routes())
export const getAppLogger = () =>
  createLogger({
    app: {
      // uhh, got config
      name: 'talkboy-9000',
      version: '2.3.6'
    }
  })
export const getAppRouter = () =>
  configureRouter(new KoaRouter<ApplicationState, IApplicationContext>())

export class App implements IApplication {
  private static instance: IApplication | undefined

  constructor(
    protected readonly koa: ApplicationKoa,
    protected readonly router: ApplicationRouter,
    protected readonly logger: winston.Logger,
    protected readonly loggingMiddleware: ILoggingMiddleware
  ) {}

  protected setup = (
    koa: ApplicationKoa,
    router: ApplicationRouter,
    logger: winston.Logger,
    loggingMiddleware: ILoggingMiddleware
  ) => {
    koa.on('error', err => {
      logger.error(err)
    })

    koa
      .use(KoaHelmet())
      .use(KoaBody())
      .use(loggingMiddleware.middleware)
      .use(router.routes())
      .use(router.allowedMethods())
  }

  listen = (port: number) => {
    this.logger.info(`listen(${port})`)
    this.setup(this.koa, this.router, this.logger, this.loggingMiddleware)

    return this.koa.listen(port)
  }

  static getInstance(): IApplication {
    if (this.instance !== undefined) return this.instance

    this.instance = new App(
      new Koa<ApplicationState, IApplicationContext>(),
      getAppRouter(),
      getAppLogger(),
      new LoggingMiddleware()
    )

    return this.instance
  }
}
