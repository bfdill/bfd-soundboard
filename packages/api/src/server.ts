import { App } from './app'

export const app = App.getInstance()

app.listen(4000)
