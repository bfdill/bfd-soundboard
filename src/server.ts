import * as Koa from 'koa'

import { KoaApp, router, logger } from './app'

export const koaApp = new KoaApp(new Koa(), router, logger)

koaApp.listen(3000)
