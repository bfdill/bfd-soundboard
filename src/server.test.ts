import { koaApp } from './server'
import { KoaApp } from './app'

describe('server', () => {
  test('has known exports', () => {
    expect(Object.keys(require('./server'))).toMatchSnapshot()
  })

  test('koaApp is KoaApp', () => {
    expect(koaApp instanceof KoaApp).toBeTruthy()
  })
})
