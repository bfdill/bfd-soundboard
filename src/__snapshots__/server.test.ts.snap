// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server has known exports 1`] = `
Array [
  "koaApp",
]
`;
